### PR TITLE
[part 2] Bypass root block loading consistency check in validation

### PR DESCRIPTION
**Pull request by ninjaahhh**
also contains cleanup on some root state methods

unlike previous PR #457, root chain is better off keeping the consistency check because it's more prone to have incomplete (inconsistent) data - we've seen in a snapshot that we have a root block in db but it lacks some of included minor blocks.  therefore using memory to keep the canonical chain seems reasonable. 

on the contrary, shard chains seem to be ok with only using memory as cache

### CODE REVIEW

File: quarkchain/cluster/root_state.py [hunk 1]
```diff
@@ -121,9 +121,9 @@ def get_root_block_by_hash(self, h, consistency_check=True):
         return RootBlock.deserialize(raw_block)
 
     def get_root_block_header_by_hash(self, h, consistency_check=True):
-        header = self.r_header_pool.get(h, None)
+        header = self.r_header_pool.get(h)
```
Comment (qcdll) [inline quarkchain/cluster/root_state.py]: this would raise Exception instead?

**Comment by qizhou** [comment]
I am not quite sure how this solves the problem of inconsistency - if the db doesn't have the data, how could we recover the cache data after restart?

File: quarkchain/cluster/root_state.py (hunk 1 above)
Comment (ninjaahhh) [reply quarkchain/cluster/root_state.py]:
```
In [1]: d = {}

In [2]: d.get(1, None) is None
Out[2]: True

In [3]: d.get(1) is None
Out[3]: True
```

Comment (ninjaahhh) [reply quarkchain/cluster/root_state.py]: `d[1]` will raise exception

**Comment by ninjaahhh** [comment]
@qizhou no it doesn't solve the inconsistency issue. if db is inconsistent, we can't magically recover missing data.

however, in root state, memory is used as a way to indicate "if I have the tip in memory, then I'm pretty sure other necessary information (e.g. included minor block header) are also recovered in memory". 

not 100% sure this kind of semantics is necessary, but this allows the tip-1 hack to work even under the inconsistent db state

**Comment by ninjaahhh** [comment]
> I am wondering, how did you identify unused imports?

all hail pycharm

**Comment by qizhou** [comment]
tip-1 trick is only used when we start a node, right?  I am still confused that since we need to recover the data from db, how cache could help?

**Comment by ninjaahhh** [comment]
> tip-1 trick is only used when we start a node, right? I am still confused that since we need to recover the data from db, how cache could help?

the memory cache holds the following invariant: **blockchain info in memory should always be consistent and complete.** the `tip-1` hack simply helps keep the invariant.

the benefits: basically the full node is more pessimistic on its data. one example - 

- after restarting, say, the node has inconsistent data on an older fork, then intentionally keeping it out of memory can force the node to re-download the fork. imagine the contrary if we don't use memory to hold the canonical chain, when querying the inconsistent fork, we will have something in storage but missing something else, thus may encounter some unexpected scenarios

**Pull request merged**